src/orion/storage/base.py

```diff
@@ -0,0 +1,126 @@
+# -*- coding: utf-8 -*-
+"""
+:mod:`orion.storage.base -- Generic Storage Protocol
+=============================================================================
+
+.. module:: base
+   :platform: Unix
+   :synopsis: Implement a generic protocol to allow Orion to communicate using
+   different storage backend
+
+"""
+from typing import Optional
+
+from orion.core.utils import Factory
+
+
+class BaseStorageProtocol:
+    """Implement a generic protocol to allow Orion to communicate using
+    different storage backend
+    """
+
+    def create_experiment(self, config):
+        """Insert a new experiment inside the database"""
+        raise NotImplementedError()
+
+    def update_experiment(self, experiment: 'Experiment', fields: Optional[dict] = None,
+                          where: Optional[dict] = None, **kwargs):
+        """Update a the fields of a given trials
+
+        Parameters
+        ----------
+        experiment: Experiment object to update
+        fields: Optional[dict] a dictionary of fields to update
+        where: constraint experiment must respect
+        kwargs: a dictionary of fields to update
+
+        Example
+        -------
+        .. code-block:: python
```

Comment:
Numpy docstring style is
```
Examples
--------
>>> query = {'status': 'RUNNING'}
>>> BaseStorageProtocol.update_experiment(experiment, fields=query)
some output ...
>>> BaseStorageProtocol.update_experiment(experiment, status='RUNNING')
```
By the way, experiments do not have status so the example is a bit misleading. ;)

Reply:
>>> make pylint fail

Reply:
What is the error message?

Reply:
experiment undefined

Reply:
Oh
```
Examples
--------
>>> experiment = Experiment('some experiment')
>>> query = {'status': 'RUNNING'}
>>> BaseStorageProtocol.update_experiment(experiment, fields=query)
some output ...
>>> BaseStorageProtocol.update_experiment(experiment, status='RUNNING')
```